### ISSUE TITLE
aur-sync: use $EDITOR instead of forcing the use of vifm

### ISSUE THREAD

**Issue by rmarquis** [AladW/aurutils#51]
It's probably not a good idea to force the use of a specific third party tool to view PKGBUILD. Instead, using the standardized `$EDITOR` variable might be preferable.

**Comment by AladW** [comment]
I went with vifm because:
+ I wanted to have all build files in one window, to avoid repeated confirmations
+ It takes directory and file names from stdin
I'm not sure how EDITOR works with that, though I guess you could implement a fallback mechanism.

**Comment by AladW** [comment]
As a compromise, I've set PAGER as a fallback when vifm is not available (9f5da3a3ae14c64c3f893856415feb32d2bdb893). It keeps in line with avoiding many "View X" prompts, and you can paste paths to another terminal.

**Comment by Earnestly** [comment]
@AladW do you think `vipe` might work as well?  If you use vim then you have access to `gf` from netrw and I believe `vipe` honours `EDITOR`.  With that you could set `EDITOR=vifm -` and get more or less the same effect.

However there may be a bug with `vipe` because even the it claims and the code seems to support it, my value for `EDITOR` is getting lost through the code, so moreutils may need to get fixed.

The theory should be sound though.

**Comment by AladW** [comment]
I'd probably still need to hardcode vifm since it requires the special argument `+view!` to enable previews, unless you split `VIEWER` on whitespace or such.

Are there some other programs that use the `VIEWER` variable?

**Comment by Earnestly** [comment]
Any "command" environment variable like `EDITOR`, `CC` will always have to suffer wordsplitting in shells, it's basically expected.  The typical solution used when spaces were an issue is to write a wrapper script and use that in, for example `CC`.

**Comment by AladW** [comment]
Another issue is that only `vifm` appears suitable if you want to actually interact with the files, rather than display their paths. I've tested the console file managers listed in archwiki with either `printf '%s\n' * | foo -` or `foo*` in my aurutils cache and got the following:

*  `ranger` which was suggested in #263 simply opens one tab for each given path
* `mc` ignores anything past the first two arguments
* `lfm` turns into a christmas tree to eventually crash
* `dired` fails with a segmentation fault
* `nffm` errors with "Inappropriate ioctl for device" and asks to delete `/boot/grub/fonts` (?!?)
* `pilot` either displays a long string or the first argument
* `netrw` (vim) opens one buffer for every argument; when taking paths from stdin, they are only printed.

Depressing. If you wanted to support these awful programs, you could symlink all relevant directories (read: all directories in the build queue) someplace and visit that directory with the file manager. Which has more problems:

* `ln -s` is slow as you need one fork per file
* `cp -rs` gives 5 million warnings due to symlinks in `.git` directories
* `lndir` ignores `.git` directories, but it can only link directories verbatim, such that false positives need to be removed by hand.

Just cloning everything to a new directory each time is another option, but then `git diff` becomes problematic. The best I can think of in that case is intertwine `aur-fetch` with the viewer... 

tl;dr unless someone has a better idea or proposes a better `lndir` I'm closing this one as wontfix.

**Comment by darkfeline** [comment]
It doesn't matter of there are other program that support this currently (which I suspect is a false premise anyway, Dired does not segfault for me), so long as the option exists, I can write my own script to support it.

All you need is some AURUTILS_BROWSER envvar set to some script/program that is called with the path to the unpacked directory as a single argument and exits when "viewing" is complete.

**Comment by AladW** [comment]
>All you need is some AURUTILS_BROWSER envvar set to some script/program that is called with the path to the unpacked directory as a single argument and exits when "viewing" is complete.

That doesn't make sense to me. There is no such thing as "the unpacked directory" since the build files required for the current process are scattered amongst all other files in `AURDEST`.

>It doesn't matter if there are other program that support this currently (which I suspect is a false premise anyway, Dired does not segfault for me),

Note that by `dired` I meant the standalone version from AUR, I haven't tried how emacs dired works in this case.

>I would prefer just the directory names either as arguments or in a file passed as the first argument.

I guess this is what you actually meant instead of the "unpacked directory" thing. I would prefer if I could use the same logic for vifm and some exposed wrapper script (the claim of "unfriendly API for stdin" is unfounded as you put it, since stdin is in the end, a file - `/dev/stdin`)

**Comment by xogium** [comment]
So, for the simplest editing, isn't it possible to just use nano ? Tbh all I use the editor for is to edit the pkgbuild itself, and very very rarely other files. I could create a synlink to nano in my own path maybe, and make aurutils use that but that doesn't sound clean nor doable. I really tried vim but I just can't get used to it... So any help would be appreciated.

**Comment by darkfeline** [comment]
@AladW The problem with using stdin to pass in the list of files is that if I am running an interactive program, there is some expectation that stdin is actually the user's input from the terminal.  However, the user cannot actually provide input to stdin because you have stolen it from the program.

This can be worked around by directly querying for the pty file for the controlling terminal, such as in my example, but it's a hack that should not be needed.

For example, if I had cause to use a command line oriented program of my choosing (which would work like ed or bash, i.e. a CLI), I would have to do the following dance:

```
#!/bin/bash
tmp=$(mktemp)
trap "rm -f $tmp" EXIT
cat >"${tmp}"
some-program </dev/$(ps hotty $$) "${tmp}"
```

**Comment by AladW** [comment]
@xogium you can set the default editor in vifm to nano in `~/.vifm/vifmrc`:
```
set vicmd=nano
```

**Comment by xogium** [comment]
Thanks a lot for that. I will just have to get used to quitting the
    file manager with :q now
    :D
    . I thought it was really enforcing vim as editor.
    
    -------- Message original --------
    
    
      @xogium you can set the default
        editor in vifm to nano in ~/.vifm/vifmrc:
      set vicmd=nano

      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view it on GitHub, or mute the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/AladW/aurutils","title":"AladW/aurutils","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/AladW/aurutils"}},"updates":{"snippets":[{"icon":"PERSON","message":"@AladW in #51: @xogium you can set the default editor in vifm to nano in `~/.vifm/vifmrc`:\r\n```\r\nset vicmd=nano\r\n```"}],"action":{"name":"View Issue","url":"https://github.com/AladW/aurutils/issues/51#issuecomment-368443945"}}}
    
    
    -- 
Ellie Reeves

**Comment by AladW** [comment]
Untested `lndir` prototype:
```
if ((view)); then
    tmp_view=$(mktemp -d)
    lndir "$tmp_view" "$PWD"
    lndir "$tmp_view" "$tmp"/patch

    # remove links to build files that are not part of the queue
    find "$tmp_view" -xtype d -maxdepth 1 | grep -Fxvf "$tmp"/queue | xargs -d '\n' rm -rf

    if [[ -v AURUTILS_VIEWER ]]; then
        command -- "$AURUTILS_VIEWER" "$tmp_view"
    else
        printf -- '%s\n' "$tmp_view"/* | command -- "${PAGER:-less}"
    fi
fi
```
Besides that it appeals to the lowest common denominator (it should even work with abominable programs like `nautilus`), the `tmp_view` directory is preserved until `aur-sync` exits. Priorly the state for the queue was lost in `vifm` simply by browsing to a different directory.

**Comment by FichteFoll** [comment]
What is the expected behavior when sending files/dirs over stdin to a differen file viewer like `vifm`?

I might try adding this feature to `ranger` when I find some time for it.

**Comment by AladW** [comment]
Personally I expect the file manager to display the files/directories from stdin, and those files only.

**Comment by AladW** [comment]
Until then I updated the lndir prototype above (it should actually work now).

**Comment by AladW** [comment]
...and updated again. Now the file manager doesn't even have to accept arguments.h

**Comment by FichteFoll** [comment]
Testing e120894 with ranger atm. Works pretty well when I filter out stuff I don't care about, but unfortunately I can't tell ranger to run a command after initialization because the way it is launched (via `command`) prevents me from doing so.

To clarify:
```
$ AUR_PAGER="ranger --cmd aur" aur sync --repo aur -u
…
/usr/lib/aurutils/aur-sync: line 327: ranger --cmd aur: command not found
```

**Comment by AladW** [comment]
For the first, I removed the quotes from AUR_PAGER so it should work as intended. You're probably right on the second, but there's not much I can do apart from writing something myself in C/C++, which is not something I had planned.

At least the patch files could be moved over to the view directory if that makes it easier to copy them.

**Comment by AladW** [comment]
Scratch that, you can just use `ln -t`
```
$ cd /tmp/tmp.crL4xFOOc8/
$ find /etc -maxdepth 1 -mindepth 1 -type d -exec ln -t "$PWD" -s {} +
```

**Comment by AladW** [comment]
New proposal:
```
if ((view)); then
    if [[ -v AUR_PAGER ]]; then
        # link build files in the queue
        find "$aur_workdir" -maxdepth 1 -mindepth 1 -type d -printf '%f\0' \
            | complement "$tmp"/queue --null | xargs -0 ln -st "$tmp_view"

        # link diffs
        ln -st "$tmp_view" "$tmp"/patch/*

        command -- $AUR_PAGER
        cd_safe "$aur_workdir"
    else
        printf '%s\n' "$tmp"/patch/* | cat - "$tmp"/queue \
            | xargs -I{} find {} -maxdepth 1 | viewer
    fi
fi
```

**Comment by AladW** [comment]
@FichteFoll https://github.com/AladW/aurutils/commit/91bd7ee889ad524d3bed0966d8f2368b24aa91a2

**Comment by FichteFoll** [comment]
```
$ AUR_PAGER="ranger --cmd aur" aur sync --repo aur dropbox
  -> Using [aur] repository
  -> dropbox: 46.4.65-2 -> 47.4.74-2
From https://aur.archlinux.org/dropbox
 = [up to date]      master     -> origin/master
mv: missing file operand
Try 'mv --help' for more information.
```

**Comment by darkfeline** [comment]
```
↪ 127 2018-04-12 20:55:21 ionasal@cielnosurge:~ $ AUR_PAGER=dired2 aur sync st
  -> Using [custom] repository
  -> st: (none) -> 0.8.1-1
Cloning into 'st'...
remote: Counting objects: 22, done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 22 (delta 9), reused 10 (delta 2)
Unpacking objects: 100% (22/22), done.
mv: missing file operand
Try 'mv --help' for more information.
```
If there are no wildcard matches, mv fails https://github.com/AladW/aurutils/commit/0013a083e8564af1d0c349cda90cfef5929bcf82#diff-09d049907497fdb78fd83e6ce9d0ec0bL324

**Comment by AladW** [comment]
https://github.com/AladW/aurutils/commit/017b70ead64a2526faf93707070b874809137e70

**Comment by AladW** [comment]
@FichteFoll if you decide to add stdin support to ranger anyway, let me know and I'll add it to the `viewer` function

**Comment by FichteFoll** [comment]
```sh
[fichte@pinaceae aurutils]$ [[ /tmp/.* ]] && echo "ok" || echo "no"
ok
[fichte@pinaceae aurutils]$ [[ /tmp/aur* ]] && echo "ok" || echo "no"
ok
[fichte@pinaceae aurutils]$ x=$(echo /tmp/.*)
[fichte@pinaceae aurutils]$ echo $x
/tmp/. /tmp/.. /tmp/.esd-1000 /tmp/.font-unix /tmp/.ICE-unix /tmp/.Test-unix /tmp/.X0-lock /tmp/.X11-unix /tmp/.XIM-unix
[fichte@pinaceae aurutils]$ [[ $x ]] && echo "ok" || echo "no"
ok
[fichte@pinaceae aurutils]$ x=$(echo /tmp/aur*)
[fichte@pinaceae aurutils]$ echo $x
[fichte@pinaceae aurutils]$ [[ $x ]] && echo "ok" || echo "no"
no
```

Looks like file expansion doesn't happen in `[[ ]]`.

**Comment by AladW** [comment]
How about https://github.com/AladW/aurutils/commit/2aafcf422b51b78a57f9bc78723c7dd4d8b96313

**Comment by FichteFoll** [comment]
Works now! (I patched it locally for testing in a similar way.)

**Comment by maximbaz** [comment]
I have an idea, with the recent changes, `vifm` opens with current directory set to `~/.cache/aurutils/sync`:

![aurutils-1](https://user-images.githubusercontent.com/1177900/39804791-5563263c-5375-11e8-9138-a1031adea3ee.png)

I think it should use the `$tmp_view` as the current directory instead:

![aurutils-2](https://user-images.githubusercontent.com/1177900/39804812-64ad838a-5375-11e8-8a1c-70d55d4c7dfa.png)

**Comment by AladW** [comment]
As something like `vifm - /tmp/view.` or `vifm /tmp/view. -` does not have the desired effect, options are:

* Specify all files on the vifm command line with a glob
  * Less robust approach as additional prompts may occur on permission errors
* `cd` to `$tmp_view` before running vifm
  * can be either done in a subshell in `viewer()` or in main if a second `cd $aur_workdir` is done afterward
* Treat `vifm` like any other file manager and just run it as `vifm $tmp_view '+view!`

However, in each of these cases files contained in these directories are not in the `vifm` prompt. This is the best you can expect for a generic file manager, but with `vifm` (and consequently, in aurutils 1.5) the path to every file was fed to `vifm -` with `find`.

**Comment by maximbaz** [comment]
I noticed that files in the directories are not immediately visible, but that's acceptable I think, as long as patches are shown directly in `$tmp_view` :)

**Comment by AladW** [comment]
1647720d00f22c76ed778389697dcc475421115c

**Comment by lepianoman** [comment]
Hi Alad, thanks so much for all the work you do with this. I hope this question isn't a duplicate: I was curious if it's possible to select packages or cancel operates from the vifm menu? Not sure how to abort besides closing the terminal (Ctrl-C doesn't work).

**Comment by Morganamilo** [comment]
`:cq`. Although I will ask the same question for less.

**Comment by franklinyu** [comment]
@Morganamilo https://unix.stackexchange.com/questions/287818/cause-less-to-exit-with-a-non-zero-status-code ?

**Comment by AladW** [comment]
The man page mentions:

>Note that use of this option makes it impossible to return to the command prompt from the "F" command.

Is that problematic to someone?

**Comment by miquecg** [comment]
Hi @AladW. I'm using a dedicated user for builds as one of the examples in the man pages shows, but I'm having troubles editing files inside vifm. No matter what key I press (`enter`, `i`) I cannot edit any file and output in the console looks like this:

> /sbin/nologin: invalid option -- 'c'
> Try 'nologin --help' for more information.

I can use vifm normally with my regular user, so the problem is either the build user or something related with sudo. Any ideas?

Thanks.

**Comment by AladW** [comment]
vifm assumes the account shell supports -c to run commands (in
particular, vim). As nologin is not a shell, it won't work. I should
probably remove --nologin from the example; feel free to open a new
issue for this.
